resources/virt.go

```diff
@@ -0,0 +1,521 @@
+package resources
+
+import (
+	"encoding/gob"
+	"log"
+	"fmt"
+	"strconv"
+
+	"github.com/purpleidea/mgmt/event"
+	"github.com/rgbkrk/libvirt-go"
+	"math/rand"
+	"time"
+	"bytes"
+)
+
+func init() {
+	gob.Register(&VirtRes{})
+}
+
+type VirtRes struct {
+	BaseRes `yaml:",inline"`
+	Cpus int `yaml:"cpus"`
+	Ram int `yaml:"ram"`
+	State string `yaml:"state"` //running, paused, shutoff
+	Transient bool `yaml:"transient"`
+	Boot []string `yaml:"boot"` // boot order. values: fd, hd, cdrom, network
+	Disk []diskDevice `yaml:"disk"`
+	CDRom []cdRomDevice `yaml:"cdrom"`
+	Network []networkDevice `yaml:"network"`
+	Filesystem []filesystemDevice `yaml:"filesystem"`
+	conn libvirt.VirConnection
+}
+
+type VirtUUID struct {
+	BaseUUID
+}
+
+type virtDevice interface {
+	GetXML(idx int) string
+}
+
+type diskDevice struct {
+	Source string `yaml:"source"`
+	Type string `yaml:"type"`
+}
+
+type cdRomDevice struct {
+	Source string `yaml:"source"`
+	Type string `yaml:"type"`
+}
+
+type networkDevice struct {
+	Name string `yaml:"name"`
+	MAC string `yaml:"mac"`
+}
+
+type filesystemDevice struct {
+	Access string `yaml:"access"`
+	Source string `yaml:"source"`
+	Target string `yaml:"target"`
+	ReadOnly bool `yaml:"read_only"`
+}
+
+var (
+	libvirtInitialized bool = false
+)
+
+func NewVirtRes(name string, cpus, ram int, state string, transient bool) *VirtRes {
+	obj := &VirtRes{
+		BaseRes: BaseRes{
+			Name: name,
+		},
+		Cpus: cpus,
+		Ram: ram,
+		State: state,
+		Transient: transient,
+	}
+	obj.Init()
+	return obj
+}
+
+func (obj *VirtRes) Init() error {
+	obj.BaseRes.kind = "Virt"
+
+	if !libvirtInitialized {
+		libvirt.EventRegisterDefaultImpl()
+		// run libvirt event loop
+		go func() {
+			for {
+				libvirt.EventRunDefaultImpl()
+			}
+		}()
+		libvirtInitialized = true
+	}
+
+	conn, err := libvirt.NewVirConnection("qemu:///system")
+	if err != nil {
+		// should we proceed or exit?
+		fmt.Println("Connection to libvirt faild")
+		return err
+	}
+	obj.conn = conn
+
+	obj.BaseRes.Init() // call base init, b/c we're overriding
+
+	return nil
+}
+
+func (obj *VirtRes) Validate() bool {
+	return true
+}
+
+func (obj *VirtRes) GetUUIDs() []ResUUID {
+	x := &VirtUUID{
+		BaseUUID: BaseUUID{name: obj.GetName(), kind: obj.Kind()},
+	}
+	return []ResUUID{x}
+}
+
+func (obj *VirtRes) Watch(processChan chan event.Event) error {
+	if obj.IsWatching() {
+		return nil
+	}
+	obj.SetWatching(true)
+	defer obj.SetWatching(false)
+	cuuid := obj.converger.Register()
+	defer cuuid.Unregister()
+
+	ch := make(chan int, 10)
```

Comment:
Does this need to be buffered for things to work? If we aren't ready to receive the event because we're still processing the last one, and there's no buffer for it does it drop? Using an unbuffered channel is probably better if we can do it. If not, please add a comment explaining why it's needed.